### ISSUE TITLE
Having merge cells populated causes 2 additional table rendering (both renderer and afterRender)

### ISSUE THREAD

**Issue by softicnermin** [handsontable/handsontable#5687]
### Description
If mergeCells is populated two additional render cycles (renderer and afterRender) are exectued. 
Is this expected behavior or is there any workaround because in cases of a bigger table (even 20x20 with some styles applied) this has significant impact on performance. For a table of 20x20 that is almost additional 800 calls.

### Steps to reproduce

1. Create a table 
2. Populate mergeCells or set it to true

### Demo
https://jsfiddle.net/w7kx2epy/

### Your environment
* Handsontable version: 6.2.2
* Browser Name and version: Chrome Version 71
* Operating System: Ubuntu 16.04

**Comment by AMBudnik** [comment]
Thank you for sharing. This subject is connected to https://github.com/handsontable/handsontable/issues/5204

**Comment by AMBudnik** [comment]
Using 8.0.0-beta1 we get the table loaded instantly while in 6.2.2 most of the cells were missing
https://jsfiddle.net/t9f4cqLr/ (8.0.0-beta1)

**Comment by AMBudnik** [comment]
We get the same behavior in v 8.1.0 https://jsfiddle.net/s2g1qud9/1/
I've simplified the demo. We can see there that enabling `mergeCells` triggers an additional log while enabling `comments` does not.

**Comment by AMBudnik** [comment]
v8 https://jsfiddle.net/mtygn76r/ (8.2.0) consoles half of those renderers. But it is still 1200 calls with merging and 400 without it.
I think that we should mention this subject when discussing partial cell rendering.

**Comment by AMBudnik** [comment]
I've tested the same scenario using `batch` method (introduced in [v8.3.0](https://handsontable.com/docs/8.3.0/tutorial-release-notes.html)) 
Demo: https://jsfiddle.net/handsoncode/z3fejcg4/

To simplify the example a bit I am loading 5 rows and 6 columns with two merged areas.
![image](https://user-images.githubusercontent.com/10757813/107937557-62453100-6f84-11eb-8919-4e4e53a3da7c.png)

Here is a difference in the numbers of calls for `afterRender` and `afterRenderer` hooks.

![image](https://user-images.githubusercontent.com/10757813/107938667-024f8a00-6f86-11eb-9205-c58039888329.png)


Not only did the batch reduced the overall number of `afterRender()` from 3 to 2 for non-merged cells and from 5 to 2 for merged areas. But it also reduced the difference between merged and non-merged data. We no longer call the extra rendering to merge the cells. And with those settings, we only call the `afterRenderer` 8 times more (in comparison to 95 times more without the batch).

**Comment by AMBudnik** [comment]
Hi @softicnermin 

I hope that the proposed batching functionality will help you to get rid of those excess calls.

Nevertheless, underneath we still need to think about improving the `mergeCells` plugin itself so I'm closing this issue and adding it to the scope of the '**Interoperability problems between MergeCells and some core features**' task reported at https://github.com/handsontable/handsontable/issues/7509

**Comment by AMBudnik** [comment]
Issue still persist. In v4 we had 1000 calls, not in 12.3.0 we have 1200 calls. Updated demo https://jsfiddle.net/ch5rn3Lg/

**Comment by PH8919** [comment]
Hi, guy
I evaluated several versions and found the versions from v8.4.0 to v11.x.x perform well for merging cells. but it degrade sharply from v12.0.0. it puzzles me. is there any reason to that and is the version v11.x.x reliable to use?

**Comment by PH8919** [comment]
@AMBudnik I shared my sample code as follows:

`<template>
  <div id="example">
    <hot-table :settings="hotSettings"></hot-table>
  </div>
</template>

<script>
import 'handsontable/dist/handsontable.full.css';

import * as Vue from 'vue';

import Handsontable from 'handsontable';

import { HotTable } from '@handsontable/vue3';

function getDebugInfo() {
  let debug = "Handsontable:";
  debug += ` v${Handsontable.version}`;
  debug += ` (${Handsontable.buildDate})`;
  debug += " Wrapper:";
  debug += ` v${HotTable.version}`;
  debug += " Vue:";
  debug += ` v${Vue.version}`;
  return debug;
}

function generateData() {
  let _data = [];
  let _mergeData = [];
  for (let i = 0; i < 10000; i++) {
    let _i = parseInt(i / 5);
    if (i % 5 == 0) {
      _mergeData.push({ row: i, col: 0, rowspan: 5, colspan: 1 });
      _mergeData.push({ row: i, col: 3, rowspan: 5, colspan: 1 });
    }
    _data.push([`A${_i}`,`B${i}`,`C${i}`,`E${i}`,`E${i}`, `A1${i}`, `B1${i}`, `C1${i}`, `E1${i}`,`E1${i}`,`A${i}`, `B${i}`, `C${i}`,`E${i}`,`E${i}`,`A1${i}`,`B1${i}`,`C1${i}`,`E1${i}`,`E1${i}`,]);
  }
  return { data: _data, mergeData: _mergeData };
}

console.log(getDebugInfo());

export default {
  name: "App",
  data: function () {
    let _ret = generateData();
    return {
      hotSettings: {
        data: _ret.data,
        colWidths:[80,80,80,80,80,80,80,80,80,80,80,80,80,80,80,80,80,80,80,80],
        colHeaders: true,
        rowHeaders: true,
        licenseKey: "non-commercial-and-evaluation",
        mergeCells: _ret.mergeData,
        height: '100%',
        width: '100%',
        dropdownMenu:['filter_by_value','filter_action_bar'],
        filters:true
      },
    };
  },

  components: {
    HotTable
  },
};
</script>`

Version info.: Handsontable: v12.4.0 (23/05/2023 09:21:52) Wrapper: v12.4.0 Vue: v3.3.4

Thanks!

**Comment by AMBudnik** [comment]
Hi @PH8919 yes, based on the code it seems that `generateData` is the function that takes the most time to execute. Especially as there's a loop for 10k elements, two merge areas added, and a data change for each interaction ot the loop.

**Comment by PH8919** [comment]
Hi @AMBudnik I also run with JS-based API in my evaluation project and append the timestamp,
`onMounted(()=>{
    let _ret = generateData();
    let div = document.getElementById('example')
    console.time('HandsonTable')
    new Handsontable(div,{
      data: _ret.data,
        colWidths:[80,80,80,80,80,80,80,80,80,80,80,80,80,80,80,80,80,80,80,80],
        colHeaders: true,
        rowHeaders: true,
        licenseKey: "non-commercial-and-evaluation",
        mergeCells: _ret.mergeData,
        height: '100%',
        width: '100%',
        dropdownMenu:['filter_by_value','filter_action_bar'],
        filters:true
    })
    console.timeEnd('HandsonTable')
  })`

it will takes about 60s to call 'new Handsontable'

Hope for your check!
Thanks!

**Comment by AMBudnik** [comment]
@PH8919 if the dataset is large it would be better to create the meta schema for the mergeCells before table creation. It seems that mering criteria can be obtains from the server before the data is loaded. So you can pass the details to the `mergeCells` as

**Comment by PH8919** [comment]
> @PH8919 if the dataset is large, it would be better to create the meta schema for the mergeCells before table creation. It seems that mering criteria can be obtained from the server before the data is loaded. So you can pass the details to the `mergeCells` as
> 
> ```js
> hot.batch(() => {
>  hot. updateSettings({mergeCells: your_configuration})
> })
> ```

@AMBudnik I followed your suggestion
` hot.batch(()=>{
      console.time('B')
      hot.updateSettings({mergeCells:_ret.mergeData})
      console.timeEnd('B')
    })`

the 'hot.updateSettings' will still take about 55s for my dataset.

Thanks!

**Comment by AMBudnik** [comment]
OK, so a slight improvement. The next step would be to optimize the 

> _data.push([`A${_i}`,`B${i}`,`C${i}`,`E${i}`,`E${i}`, `A1${i}`, `B1${i}`, `C1${i}`, `E1${i}`,`E1${i}`,`A${i}`, `B${i}`, `C${i}`,`E${i}`,`E${i}`,`A1${i}`,`B1${i}`,`C1${i}`,`E1${i}`,`E1${i}`,]); }

Could you also do that outside the `onMounted`?

**Comment by PH8919** [comment]
@AMBudnik I must tell you that when I go back to v11.x.x with "handsontable": "^11.0.0", it only takes about 4s to build the table with "new handsonTable". with the same test code and test data.

Thanks!

**Comment by AMBudnik** [comment]
Hm... Maybe you could send us a full demo for tests via support@handsontable.com, @PH8919 ?
That should bring the best results.

**Comment by PH8919** [comment]
Hi @AMBudnik I am sorry I cannot share my demo with e-mail due to some reason, I can only paste it here and you can try it. it belongs to App.vue in vue demo project. 
`<template>
  <div id="example" style="width: 100%;height: 100%;">
    
  </div>
</template>

<script>
import 'handsontable/dist/handsontable.full.css';

import * as Vue from 'vue';

import Handsontable from 'handsontable';

import { HotTable } from '@handsontable/vue3';

function getDebugInfo() {
  let debug = "Handsontable:";
  debug += ` v${Handsontable.version}`;
  debug += ` (${Handsontable.buildDate})`;
  debug += " Wrapper:";
  debug += ` v${HotTable.version}`;
  debug += " Vue:";
  debug += ` v${Vue.version}`;
  return debug;
}

function generateData() {
  let _data = [];
  let _mergeData = [];
  for (let i = 0; i < 10000; i++) {
    let _i = parseInt(i / 5);
    if (i % 5 == 0) {
      _mergeData.push({ row: i, col: 0, rowspan: 5, colspan: 1 });
      _mergeData.push({ row: i, col: 3, rowspan: 5, colspan: 1 });
    }
    _data.push([`A${_i}`,`B${i}`,`C${i}`,`E${i}`,`E${i}`, `A1${i}`, `B1${i}`, `C1${i}`, `E1${i}`,`E1${i}`,`A${i}`, `B${i}`, `C${i}`,`E${i}`,`E${i}`,`A1${i}`,`B1${i}`,`C1${i}`,`E1${i}`,`E1${i}`,]);
  }
  return { data: _data, mergeData: _mergeData };
}

console.log(getDebugInfo());
let _ret = generateData();
export default {
  name: "App",
  data: function () {
    return {
    };
  },
  setup(){
    Vue.onMounted(()=>{
    
    let div = document.getElementById('example')
    console.time('HandsonTable')
    let table = new Handsontable(div,{
      data: _ret.data,
        colWidths:[80,80,80,80,80,80,80,80,80,80,80,80,80,80,80,80,80,80,80,80],
        colHeaders: false,
        rowHeaders: true,
        licenseKey: "non-commercial-and-evaluation",
        // mergeCells: _ret.mergeData,
        height: '100%',
        width: '100%',
        // dropdownMenu:['filter_by_value','filter_action_bar'],
        dropdownMenu: {
          items: {
            filter_by_value: {
              // hide the 'Filter by value' list from all columns but the first one
              hidden() {
                return this.getSelectedRangeLast().to.col > 0;
              },
            },
            filter_action_bar: {
              // hide the 'OK' and 'Cancel' buttons from all columns but the first one
              hidden() {
                return this.getSelectedRangeLast().to.col > 0;
              },
            }}
        },
        filters:true,
        afterGetColHeader(col, TH) {
    // remove the column menu button from the 'Brand', 'Price', and 'Date' columns
    if (col > 0) {
      const button = TH.querySelector('.changeType');

      if (!button) {
        return;
      }

      button.parentElement.removeChild(button);
    }
  }})
    console.timeEnd('HandsonTable')
  
    table.batch(()=>{
      console.time('B')
      table.updateSettings({mergeCells:_ret.mergeData})
      console.timeEnd('B')
    })
  })
  }
};
</script>

<style>
body {
  height: 100%;
  width: 100%;
  display: block;
}
</style>
`

**Comment by AMBudnik** [comment]
Thank you additional details, @PH8919 I see that you posted on the forum so we can continue there.

**Comment by PH8919** [comment]
Hi @AMBudnik Okay. that is a good suggestion.

**Comment by budnix** [comment]
Hi @softicnermin

I re-tested this on the latest version (18.1.0) and the problem is still there, so I have now fixed it.

I counted the draws on a 10x10 grid, where one full draw is exactly 100 renderer calls:

| Setup | Draws before | Draws after |
| --- | --- | --- |
| no `mergeCells` | 1 | 1 |
| `mergeCells: true` | 2 | 1 |
| `mergeCells` with merged areas | 3 | 2 |

Both extra draws came from the same place. The plugin applies the merges only after the grid has already been drawn once. Clearing the cells that a merge covers caused one draw, and the plugin then asked for another one straight after.

The fix groups that work into a single draw, and skips it completely when there is nothing to merge. A grid with `mergeCells: true` now costs no extra draw at all, and a grid with merged areas costs one instead of two.

One extra draw remains. The plugin can only apply the merges after the first draw has happened, so removing that last one needs a bigger change to the startup order. That is out of scope here.

You also no longer need the `batch()` workaround mentioned earlier in this thread when you pass `mergeCells` in the constructor.

The fix is in #13362 and will ship in one of the upcoming releases.
